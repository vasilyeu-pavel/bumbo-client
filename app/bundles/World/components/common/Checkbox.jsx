import React, { Component, PropTypes } from 'react';
import { bindAll, map } from 'lodash';
import classNames from 'classnames';

export default class Checkbox extends Component {
  static propTypes = {
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    checked: PropTypes.bool,
    labelText: PropTypes.string,
    onChange: PropTypes.func,
    modifiers: PropTypes.array,
  }

  static defaultProps = {
    checked: false,
    modifiers: [],
  }

  constructor(props, context) {
    super(props, context);

    this.state = { checked: this.props.checked };

    bindAll(this, 'handleChange');
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.checked !== this.props.checked) {
      this.setState({ checked: nextProps.checked });
    }
  }

  handleChange(event) {
    this.props.onChange(this.props.id, !this.state.checked);
    this.setState({ checked: !this.state.checked });
  }

  getClassNames() {
    return classNames('checkbox', map(this.props.modifiers,
      (modifier) => `checkbox--${modifier}`));
  }

  render() {
    return (
      <div className={this.getClassNames()}>
        <input
          ref={'checkbox'}
          className='checkbox__real'
          id={this.props.id}
          type='checkbox'
          checked={this.state.checked}
          onChange={this.handleChange}
        />
        <label className='checkbox__label' htmlFor={this.props.id}>
          <div className='checkbox__icon'></div>
          <div className='checkbox__label-text'>
            {this.props.labelText}
            <div className='checkbox__label-caption'>{this.props.captionText}</div>
          </div>
        </label>
      </div>
    );
  }
}
