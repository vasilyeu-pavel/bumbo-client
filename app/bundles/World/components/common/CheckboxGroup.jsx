import React, { Component, PropTypes } from 'react';
import { bindAll, includes, pull, clone, map } from 'lodash';
import Checkbox from './Checkbox.jsx';

export default class CheckboxGroup extends Component {
  static propTypes = {
    name: PropTypes.string,
    subtitle: PropTypes.string,
    values: PropTypes.oneOfType([PropTypes.array, PropTypes.object]).isRequired,
    checked: PropTypes.array.isRequired,
    onChange: PropTypes.func.isRequired,
  }

  constructor(props, context) {
    super(props, context);

    this.checkedItems = clone(props.checked) || [];

    bindAll(this, ['handleChange']);
  }

  handleChange(id, isChecked) {
    if (isChecked && !includes(this.checkedItems, id)) {
      this.checkedItems.push(id);
    } else {
      pull(this.checkedItems, id);
    }

    this.props.onChange(this.props.name, this.checkedItems);
  }

  render() {
    return (
      <div className='checkbox-group'>
        <div className='checkbox-group__subtitle'>{this.props.subtitle}</div>
        {
          map(this.props.values, (key, value) => (
            <div key={key} className='checkbox-group__item'>
              <Checkbox
                id={key}
                checked={includes(this.props.checked, key)}
                labelText={value}
                onChange={this.handleChange}
              />
            </div>
          ))
        }
      </div>
    );
  }
}
