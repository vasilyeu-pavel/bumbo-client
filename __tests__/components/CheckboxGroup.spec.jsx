/* global expect, jest */
/* eslint arrow-body-style: 0 */

import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import CheckboxGroup from '../../app/bundles/World/components/common/CheckboxGroup';
import Checkbox from '../../app/bundles/World/components/common/Checkbox';

describe('<CheckboxGroup /> component', () => {
  const minProps = {
    name: 'beachTypes',
    subtitle: 'Beach types',
    values: { 'Песок белый': 151, 'Песок вулканический': 149, 'Песок жёлтый': 150 },
    checked: [],
    onChange: () => {},
  };
  const valuesCount = Object.keys(minProps.values).length;

  it(`should render ${valuesCount} Checkbox'es`, () => {
    const checkboxGroup = renderer.create(<CheckboxGroup {...minProps} />);
    const checkboxGroupTree = checkboxGroup.toJSON();

    expect(checkboxGroupTree).toMatchSnapshot();
  });

  it(`should render ${valuesCount} Checkbox component instances`, () => {
    const checkboxGroup = shallow(<CheckboxGroup {...minProps} />);

    checkboxGroup.find('.checkbox-group__item').forEach((node) => {
      expect(node.children().shallow().instance()).toBeInstanceOf(Checkbox);
    });
  });

  it(`should call "onChange" method with name ${minProps.name}`, () => {
    const handleChange = (name) => expect(name).toEqual(minProps.name);
    const handler = jest.fn(handleChange);
    const checkboxGroup = renderer.create(<CheckboxGroup {...minProps} onChange={handler} />);
    let checkboxGroupTree = checkboxGroup.toJSON();
    const checkboxes = checkboxGroupTree.children;
    const checkbox = checkboxes
      .filter(child => child.props.className === 'checkbox-group__item')[0];
    const input = checkbox.children[0].children.filter(child => child.type === 'input')[0];

    // Click on first checkbox
    input.props.onChange();

    expect(handler).toHaveBeenCalled();

    checkboxGroupTree = checkboxGroup.toJSON();

    expect(checkboxGroupTree).toMatchSnapshot();

    // Second click on first checkbox
    input.props.onChange();

    expect(handler).toHaveBeenCalledTimes(2);

    checkboxGroupTree = checkboxGroup.toJSON();

    expect(checkboxGroupTree).toMatchSnapshot();
  });

  it('should add an id to "checked" array after click on checkbox', () => {
    const handleChange = (name, checkedArray) => expect(checkedArray).toEqual([151]);
    const handler = jest.fn(handleChange);
    const checkboxGroup = mount(<CheckboxGroup {...minProps} onChange={handler} />);

    checkboxGroup.find('input[type="checkbox"]').first().simulate('change');
    expect(handler).toHaveBeenCalled();
  });
});
