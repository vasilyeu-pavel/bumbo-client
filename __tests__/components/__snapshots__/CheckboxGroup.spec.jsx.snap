exports[`<CheckboxGroup /> component should call "onChange" method with name beachTypes 1`] = `
<div
  className="checkbox-group">
  <div
    className="checkbox-group__subtitle">
    Beach types
  </div>
  <div
    className="checkbox-group__item">
    <div
      className="checkbox ">
      <input
        checked={true}
        className="checkbox__real"
        id={151}
        onChange={[Function]}
        type="checkbox" />
      <label
        className="checkbox__label"
        htmlFor={151}>
        <div
          className="checkbox__icon" />
        <div
          className="checkbox__label-text">
          Песок белый
        </div>
      </label>
    </div>
  </div>
  <div
    className="checkbox-group__item">
    <div
      className="checkbox ">
      <input
        checked={false}
        className="checkbox__real"
        id={149}
        onChange={[Function]}
        type="checkbox" />
      <label
        className="checkbox__label"
        htmlFor={149}>
        <div
          className="checkbox__icon" />
        <div
          className="checkbox__label-text">
          Песок вулканический
        </div>
      </label>
    </div>
  </div>
  <div
    className="checkbox-group__item">
    <div
      className="checkbox ">
      <input
        checked={false}
        className="checkbox__real"
        id={150}
        onChange={[Function]}
        type="checkbox" />
      <label
        className="checkbox__label"
        htmlFor={150}>
        <div
          className="checkbox__icon" />
        <div
          className="checkbox__label-text">
          Песок жёлтый
        </div>
      </label>
    </div>
  </div>
</div>
`;

exports[`<CheckboxGroup /> component should call "onChange" method with name beachTypes 2`] = `
<div
  className="checkbox-group">
  <div
    className="checkbox-group__subtitle">
    Beach types
  </div>
  <div
    className="checkbox-group__item">
    <div
      className="checkbox ">
      <input
        checked={false}
        className="checkbox__real"
        id={151}
        onChange={[Function]}
        type="checkbox" />
      <label
        className="checkbox__label"
        htmlFor={151}>
        <div
          className="checkbox__icon" />
        <div
          className="checkbox__label-text">
          Песок белый
        </div>
      </label>
    </div>
  </div>
  <div
    className="checkbox-group__item">
    <div
      className="checkbox ">
      <input
        checked={false}
        className="checkbox__real"
        id={149}
        onChange={[Function]}
        type="checkbox" />
      <label
        className="checkbox__label"
        htmlFor={149}>
        <div
          className="checkbox__icon" />
        <div
          className="checkbox__label-text">
          Песок вулканический
        </div>
      </label>
    </div>
  </div>
  <div
    className="checkbox-group__item">
    <div
      className="checkbox ">
      <input
        checked={false}
        className="checkbox__real"
        id={150}
        onChange={[Function]}
        type="checkbox" />
      <label
        className="checkbox__label"
        htmlFor={150}>
        <div
          className="checkbox__icon" />
        <div
          className="checkbox__label-text">
          Песок жёлтый
        </div>
      </label>
    </div>
  </div>
</div>
`;

exports[`<CheckboxGroup /> component should render 3 Checkbox'es 1`] = `
<div
  className="checkbox-group">
  <div
    className="checkbox-group__subtitle">
    Beach types
  </div>
  <div
    className="checkbox-group__item">
    <div
      className="checkbox ">
      <input
        checked={false}
        className="checkbox__real"
        id={151}
        onChange={[Function]}
        type="checkbox" />
      <label
        className="checkbox__label"
        htmlFor={151}>
        <div
          className="checkbox__icon" />
        <div
          className="checkbox__label-text">
          Песок белый
        </div>
      </label>
    </div>
  </div>
  <div
    className="checkbox-group__item">
    <div
      className="checkbox ">
      <input
        checked={false}
        className="checkbox__real"
        id={149}
        onChange={[Function]}
        type="checkbox" />
      <label
        className="checkbox__label"
        htmlFor={149}>
        <div
          className="checkbox__icon" />
        <div
          className="checkbox__label-text">
          Песок вулканический
        </div>
      </label>
    </div>
  </div>
  <div
    className="checkbox-group__item">
    <div
      className="checkbox ">
      <input
        checked={false}
        className="checkbox__real"
        id={150}
        onChange={[Function]}
        type="checkbox" />
      <label
        className="checkbox__label"
        htmlFor={150}>
        <div
          className="checkbox__icon" />
        <div
          className="checkbox__label-text">
          Песок жёлтый
        </div>
      </label>
    </div>
  </div>
</div>
`;
